src/algebra/divisibility.lean

```diff
@@ -89,4 +128,67 @@ dvd.elim h (assume c, assume H' : a = 0 * c, eq.trans H' (zero_mul c))
 
 @[simp] theorem dvd_zero (a : α) : a ∣ 0 := dvd.intro 0 (by simp)
 
-end comm_monoid_with_zero
+end monoid_with_zero
+
+/-- Given two elements b, c of an integral domain and a nonzero element a, `a*b` divides `a*c` iff
+  b divides c. -/
+theorem mul_dvd_mul_iff_left [cancel_monoid_with_zero α] {a b c : α}
+  (ha : a ≠ 0) : a * b ∣ a * c ↔ b ∣ c :=
+exists_congr $ λ d, by rw [mul_assoc, mul_right_inj' ha]
+
+/-!
+### Units in various monoids
+-/
+
+namespace units
+
+section monoid
+variables [monoid α] (a b : α) (u : units α)
+
+/-- Elements of the unit group of a monoid represented as elements of the monoid
+    divide any element of the monoid. -/
+@[simp] lemma coe_dvd : ↑u ∣ a := ⟨↑u⁻¹ * a, by simp⟩
+
+/-- In a monoid, an element `a` divides an element `b` iff `a` divides all
+    associates of `b`. -/
+@[simp] lemma dvd_coe_mul : a ∣ b * u ↔ a ∣ b :=
+iff.intro
+  (assume ⟨c, eq⟩, ⟨c * ↑u⁻¹, by rw [← mul_assoc, ← eq, units.mul_inv_cancel_right]⟩)
+  (assume ⟨c, eq⟩, eq.symm ▸ dvd_mul_of_dvd_left (dvd_mul_right _ _) _)
+
+/-- An element of a monoid divides a unit iff the element divides one. -/
+@[simp] lemma dvd_coe : a ∣ ↑u ↔ a ∣ 1 :=
+suffices a ∣ 1 * ↑u ↔ a ∣ 1, by simpa,
+dvd_coe_mul _ _ _
+
+end monoid
+
+/-- In a monoid, an element a divides an element b iff all associates of a divide b.-/
```

Comment:
```suggestion
/-- In a commutative monoid, an element `a` divides an element `b` iff all associates of `a` divide `b`.-/
```

Reply:
I made this one work for noncommutative monoids.